File: THUCHANH/Lab06/Exercise02/GitHubOAuth_demo.java
import java.io.*;
import java.net.*;
import java.util.*;

public class GitHubOAuth_demo {
    private static final String CLIENT_ID = "Ov23ligKsdM95J3vo1fi";
    private static final String CLIENT_SECRET = "996d8a6b3b7dd0023bd231db8598541a4dd1f530";
    private static final String AUTH_ENDPOINT = "https://github.com/login/oauth/authorize?response_type=code&client_id=" + CLIENT_ID;
    private static final String TOKEN_ENDPOINT = "https://github.com/login/oauth/access_token";
    private static final String USER_ENDPOINT = "https://api.github.com/user";

    public static void main(String[] args) throws IOException {
        // Open GitHub authorization page
        System.out.println("Opening browser for GitHub authorization...");
        Desktop.getDesktop().browse(new URI(AUTH_ENDPOINT));

        // Get authorization code from user
        System.out.print("Enter the auth code: ");
        Scanner scanner = new Scanner(System.in);
        String authCode = scanner.nextLine();

        // Exchange auth code for access token
        String accessToken = getAccessToken(authCode);
        
        // Fetch user data with access token
        if (accessToken != null) {
            fetchUserData(accessToken);
        }
    }

    private static String getAccessToken(String authCode) throws IOException {
        URL url = new URL(TOKEN_ENDPOINT);
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setRequestMethod("POST");
        connection.setRequestProperty("Accept", "application/json");
        connection.setDoOutput(true);

        String data = "client_id=" + CLIENT_ID + "&client_secret=" + CLIENT_SECRET + "&code=" + authCode;
        try (OutputStream os = connection.getOutputStream()) {
            byte[] input = data.getBytes("utf-8");
            os.write(input, 0, input.length);
        }

        BufferedReader in = new BufferedReader(new InputStreamReader(connection.getInputStream(), "utf-8"));
        StringBuilder response = new StringBuilder();
        String line;

        while ((line = in.readLine()) != null) {
            response.append(line);
        }
        
        in.close();

        // Parse the response to extract the access token
        String accessToken = null;
        String[] params = response.toString().split("&");
        for (String param : params) {
            if (param.startsWith("access_token=")) {
                accessToken = param.split("=")[1];
                break;
            }
        }
        
        return accessToken;
    }

    private static void fetchUserData(String token) throws IOException {
        URL url = new URL(USER_ENDPOINT);
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setRequestProperty("Authorization", "token " + token);
        connection.setRequestProperty("Accept", "application/json");

        BufferedReader in = new BufferedReader(new InputStreamReader(connection.getInputStream(), "utf-8"));
        StringBuilder response = new StringBuilder();
        String line;

        while ((line = in.readLine()) != null) {
            response.append(line);
        }

        in.close();

        // Print user information (you may need to parse JSON in a structured way)
        System.out.println("User Data: " + response.toString());
    }
}

File: THUCHANH/Lab06/Exercise02/GitHubOAuth_demo.cs
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading.Tasks;
using System.Collections.Generic;

class GitHubOAuth
{
    private static readonly string CLIENT_ID = "Ov23ligKsdM95J3vo1fi";
    private static readonly string CLIENT_SECRET = "996d8a6b3b7dd0023bd231db8598541a4dd1f530";
    private static readonly string AUTH_ENDPOINT = $"https://github.com/login/oauth/authorize?response_type=code&client_id={CLIENT_ID}";
    private static readonly string TOKEN_ENDPOINT = "https://github.com/login/oauth/access_token";
    private static readonly string USER_ENDPOINT = "https://api.github.com/user";

    static async Task Main(string[] args)
    {
        // Open authorization page
        Console.WriteLine("Opening browser for GitHub authorization...");
        System.Diagnostics.Process.Start(new System.Diagnostics.ProcessStartInfo
        {
            FileName = AUTH_ENDPOINT,
            UseShellExecute = true
        });

        // Get authorization code from user
        Console.Write("Enter the auth code: ");
        string authCode = Console.ReadLine();

        // Get access token
        string accessToken = await GetAccessToken(authCode);
        if (accessToken != null)
        {
            await FetchUserData(accessToken);
        }
    }

    private static async Task<string> GetAccessToken(string authCode)
    {
        using (var client = new HttpClient())
        {
            var content = new FormUrlEncodedContent(new[]
            {
                new KeyValuePair<string, string>("client_id", CLIENT_ID),
                new KeyValuePair<string, string>("client_secret", CLIENT_SECRET),
                new KeyValuePair<string, string>("code", authCode)
            });

            var response = await client.PostAsync(TOKEN_ENDPOINT, content);
            var responseString = await response.Content.ReadAsStringAsync();

            // Extract access token from the response
            var query = System.Web.HttpUtility.ParseQueryString(responseString);
            return query["access_token"];
        }
    }

    private static async Task FetchUserData(string token)
    {
        using (var client = new HttpClient())
        {
            client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
            client.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("AppName", "1.0")); // Required by GitHub API

            var response = await client.GetAsync(USER_ENDPOINT);
            var userData = await response.Content.ReadAsStringAsync();

            Console.WriteLine("User Data: " + userData);
        }
    }
}
